### DISCUSSION TITLE
JPA Entity에는 왜 기본 생성자가 필요할까요?

### DISCUSSION THREAD

**Discussion by Kimprodp** [2024-woowacourse-study/level-interview#163] [Spring Data]
proxy 객체

**Comment by alstn113** [comment]
JPA 구현체는 엔티티를 데이터베이스에서 조회해 올 때 리플렉션을 사용하여 객체를 인스턴스화합니다.
왜냐하면 JPA는 우리가 엔티티로 어떤 타입을 생성할지 알 수 없기 때문입니다.
리플렉션은 클래스의 기본 생성자를 호출하여 객체를 생성하는 방식을 사용하는데, 따라서 엔티티에는 기본 생성자가 필요합니다.

또한 프록시와 지연 로딩때문에 기본 생성자가 private이면 안됩니다.
JPA는 연관된 엔티티를 지연 로딩으로 처리할 때 프록시 객체를 사용합니다. 
프록시 객체는 실제 엔티티를 상속받아 만들어지며, 이를 위해 기본 생성자가 필요합니다. 
만약 기본 생성자가 private이라면 프록시 객체를 생성할 수 없기 때문에 지연 로딩이 동작하지 않습니다.

참고: https://prolog.techcourse.co.kr/studylogs/2489

**Comment by 3Juhwan** [comment]
JPA 엔티티 클래스는 기본 생성자를 필요로 하는 이유는 다음과 같습니다. 

**리플렉션 사용**
JPA는 리플렉션을 사용하여 엔티티 객체를 생성합니다. 리플렉션을 통해 클래스의 객체를 생성할 때 기본 생성자가 필요합니다. 

**프록시 객체 생성을 위해**
JPA는 지연 로딩과 같은 기능을 제공하기 위해 프록시 객체를 생성할 때 기본 생성자를 사용합니다. 프록시 객체는 데이터베이스에 실제로 접근할 필요가 있을 때까지 데이터를 로드하지 않는 객체입니다. 이 프록시 객체를 생성할 때 JPA는 기본 생성자를 통해 객체를 생성하고 나중에 필요한 데이터를 채워 넣습니다.

**JPA 표준 준수**
JPA 사양에서는 엔티티 클래스가 public 또는 protected의 기본 생성자를 가져야 한다고 명시하고 있습니다. 이는 JPA 구현체들이 일관된 방법으로 엔티티 객체를 생성할 수 있도록 보장하기 위함입니다.

추가로 기본 생성자의 접근 제어자가 private이어도 컴파일 타임에 예외가 발생하지 않습니다. 프록시 객체를 사용할 일이 없다면 애플리케이션은 정상 동작합니다. 

참고로 리플렉션을 통해 기본 생성자의 파라미터 정보를 받아올 수 없다는 것은 부정확한 정보입니다. [여기](https://colour-my-memories-blue.tistory.com/16)를 참고해 주세요. 

**참고 자료**

https://docs.jboss.org/hibernate/orm/3.5/reference/en/html/persistent-classes.html#persistent-classes-pojo-constructor
https://prolog.techcourse.co.kr/studylogs/2489
